motoman_driver/MotoPlus/IoServer.h

```diff
@@ -34,10 +34,240 @@
 #define IOSERVER_H
 
 extern void Ros_IoServer_StartNewConnection(Controller* controller, int sd);
+extern void Ros_IoServer_StopConnection(Controller* controller, int connectionIndex);
+
+extern void Ros_IoServer_WaitForSimpleMsg(Controller* controller, int connectionIndex);
+extern BOOL Ros_IoServer_SimpleMsgProcess(SimpleMsg* receiveMsg, SimpleMsg* replyMsg);
 
 extern int Ros_IoServer_ReadIOBit(SimpleMsg* receiveMsg, SimpleMsg* replyMsg);
 extern int Ros_IoServer_WriteIOBit(SimpleMsg* receiveMsg, SimpleMsg* replyMsg);
 extern int Ros_IoServer_ReadIOGroup(SimpleMsg* receiveMsg, SimpleMsg* replyMsg);
 extern int Ros_IoServer_WriteIOGroup(SimpleMsg* receiveMsg, SimpleMsg* replyMsg);
+extern int Ros_IoServer_ReadIORegister(SimpleMsg* receiveMsg, SimpleMsg* replyMsg);
+extern int Ros_IoServer_WriteIORegister(SimpleMsg* receiveMsg, SimpleMsg* replyMsg);
+
+typedef enum
+{
+	IO_ACCESS_BIT,
+	IO_ACCESS_GROUP,
+	IO_ACCESS_REGISTER
+} IoAccessSize;
+
+extern BOOL Ros_IoServer_IsValidReadAddress(UINT32 address, IoAccessSize size);
+extern BOOL Ros_IoServer_IsValidWriteAddress(UINT32 address, IoAccessSize size);
+extern BOOL Ros_IoServer_IsValidWriteValue(UINT32 value, IoAccessSize size);
+
+//**********************************************************
+#if DX100
```

Comment:
These constants seem like they overlap in many places.

Would it be worth it to reuse them if they have the same values? Or would you want to keep the duplicated ones?

Reply:
I don't understand your comment. Can you elaborate?

If you're referring the usage of the controller-model constant, this is something that is defined in the build system for the MotoPlus SDK.

Reply:
Example: `SPECIFICINMIN` and `SPECIFICINMAX` seem to be the same across all supported controller models.

Instead of defining the values 5 times, we could also do it once, and only provide `#define`s for the values which are different.

Reply:
Yeah, I originally tried to determine what all the differences were. But, it started to get confusing and difficult to read the code due to all the similarities/differences between controllers.

I think it's easier to read if they are all grouped nicely in a block in numerical order. It will also be easier to update in the future, if this code is around for the next controller generation.

Reply:
Ok. Makes sense.